waku/v2/protocol/waku_noise/noise.nim

```diff
@@ -246,4 +254,143 @@ proc decryptNoisePublicKey*(cs: ChaChaPolyCipherState, noisePublicKey: NoisePubl
   else:
     decryptedNoisePublicKey = noisePublicKey
-  return decryptedNoisePublicKey
\ No newline at end of file
+  return decryptedNoisePublicKey
+
+
+#################################################################
+
+# Payload encoding/decoding procedures
+
+# Checks equality between two PayloadsV2 objects
+proc `==`(p1, p2: PayloadV2): bool =
+  return (p1.protocolId == p2.protocolId) and 
+         (p1.handshakeMessage == p2.handshakeMessage) and 
+         (p1.transportMessage == p2.transportMessage) 
+  
+
+# Generates a random PayloadV2
+proc randomPayloadV2*(rng: var BrHmacDrbgContext): PayloadV2 =
+  var payload2: PayloadV2
+  # To generate a random protocol id, we generate a random 1-byte long sequence, and we convert the first element to uint8
+  payload2.protocolId = randomSeqByte(rng, 1)[0].uint8
+  # We set the handshake message to three unencrypted random Noise Public Keys
+  payload2.handshakeMessage = @[genNoisePublicKey(rng), genNoisePublicKey(rng), genNoisePublicKey(rng)]
+  # We set the transport message to a random 128-bytes long sequence
+  payload2.transportMessage = randomSeqByte(rng, 128)
+  return payload2
+
+
+# Serializes a PayloadV2 object to a byte sequences according to https://rfc.vac.dev/spec/35/.
+# The results can be passed to the payload field of a WakuMessage https://rfc.vac.dev/spec/14/
+proc serializePayloadV2*(self: PayloadV2): Result[seq[byte], cstring] 
+  {.raises: [Defect, NoiseMalformedHandshake, NoisePublicKeyError].} =
+
+  #We collect public keys contained in the handshake message
+  var
+    # According to https://rfc.vac.dev/spec/35/, the maximum size for the handshake message is 256 bytes, that is 
+    # the handshake message length can be represented with 1 byte only. (its length can be stored in 1 byte)
+    # However, to ease public keys length addition operation, we declare it as int and later cast to uit8
+    serializedHandshakeMessageLen: int = 0
+    # This variables will store the concatenation of the serializations of all public keys in the handshake message
+    serializedHandshakeMessage = newSeqOfCap[byte](256)
+    # A variable to store the currently processed public key serialization
+    serializedPk: seq[byte]
+  # For each public key in the handshake message
+  for pk in self.handshakeMessage:
+    # We serialize the public key
+    serializedPk = serializeNoisePublicKey(pk)
+    # We sum its serialized length to the total
+    serializedHandshakeMessageLen +=  serializedPk.len
+    # We add its serialization to the concatenation of all serialized public keys in the handshake message 
+    serializedHandshakeMessage.add serializedPk
+    # If we are processing more than 256 byte, we return an error
+    if serializedHandshakeMessageLen > 256:
+      debug "PayloadV2 malformed: too many public keys contained in the handshake message"
+      raise newException(NoiseMalformedHandshake, "Too many public keys in handshake message")
```

Comment:
Addressed in https://github.com/status-im/nim-waku/commit/430f2ecd6720d5835c430e310ffb06ce27f840c7.

However, would it better to keep track of specific returned error objects, e.g. `NoisePublicKeyError`, `NoiseMalformedHandshake` etc. ? It seems that with `err`, functions which call a proc with a `Result` output, will loose the information regarding the type of errors returned by the called proc.

Reply:
Good question, as per Nim style guide, [Error handling section](https://status-im.github.io/nim-style-guide/errors.html#error-handling) 
> Because of the controversies and changing landscape, the preference for Status projects is to avoid the use of exceptions unless specially motivated, if only to maintain consistency and simplicity.